### PR TITLE
Simplify endian detection logic

### PR DESCRIPTION
**Pull request by justinchuby**
Simplify endian detection logic by removing if branches and leveraging c++ syntax.

Reference: https://github.com/pytorch/pytorch/blob/653c1564bf4c46d9758000ea70a610ee3699acf1/caffe2/operators/fused_rowwise_nbitfake_conversion_ops.h#L16-L19

### Motivation and Context

Code quality

### CODE REVIEW

File: onnx/common/platform_helpers.h [hunk 1]
```diff
@@ -6,14 +6,14 @@
...
 namespace ONNX_NAMESPACE {
 
 // Determine if the processor is little endian or not
 inline bool is_processor_little_endian() {
-  int num = 1;
-  if (*(char*)&num == 1)
-    return true;
-  return false;
+  constexpr std::int32_t value = 1;
+  return reinterpret_cast<const std::uint8_t*>(&value)[0] == 1;
```
Comment (jcwchen) [inline onnx/common/platform_helpers.h:16]: Wouldn't `return *(char*)&num == 1;` be more simplified? I am fine with either way tho.

**Review by jcwchen — APPROVED** [review]

**Review by gramalingam — APPROVED** [review]

File: onnx/common/platform_helpers.h (hunk 1 above)
Comment (justinchuby) [reply onnx/common/platform_helpers.h:16]: I understand they are really the same thing, but reinterpret_cast etc. is c++ syntax to make things more explicit. I can look at the disassembled code and compare

Comment (jcwchen) [reply onnx/common/platform_helpers.h:16]: No worries. No need to spend more time on it. Feel free to forward this PR. Thanks!

**Comment by justinchuby** [comment]
Assembly is exactly the same:

image

**Pull request merged**